src/s2and_ext/clustering_models.py

```diff
@@ -0,0 +1,167 @@
+        """
+        Calculates distance matrices for each block in the block_dict
+        """
+        block_to_dmatrix = {}
+        for block_name in tqdm(block_dict):
+            block = block_dict[block_name]
+            dmatrix = self.get_distance_matrix(block)
+            block_to_dmatrix[block_name] = dmatrix
+        return block_to_dmatrix
+
+    def predict(
+        self, 
+        block_dict: Dict[str, List[str]], 
+        block_to_dmatrix: Dict[str, np.ndarray] = None
+    ) -> Dict[str, str]:
+        """
+        Predicts clusters for signatures in block_dict
+        """
+        sign_to_pred_clusters = {}
+        for block_name in block_dict:
+            block = block_dict[block_name]
+            if block_to_dmatrix is None:
+                dmatrix = self.get_distance_matrix(block)
+            else:
+                dmatrix = block_to_dmatrix[block_name]
+            # Resolve issue with dmatrix of 1 element in agglomerative clustering
+            if self.clusterer_name == 'agglomerative' and dmatrix.shape == (1,1):
+                clusters=[0]
+            else:
+                clusters = self.clusterer.fit_predict(dmatrix)
+
+            max_identifier = max(clusters)
+            for signature, cluster in zip(block, clusters):
+                
+                if cluster == -1:
+                    max_identifier += 1
+                    sign_to_pred_clusters[signature] = block_name + '_' + str(max_identifier)
+                else:
+                    sign_to_pred_clusters[signature] = block_name + '_' + str(cluster)  
+        return sign_to_pred_clusters
+
+class DummyClusterer():
+    """
+    Dummy class used due to compatibilty issues with cluster_eval function 
+    which is provided by S2AND. Must implement predict method that maps
+    cluster names to their signatures
+    """
+
+    def __init__(self, source) -> None:
+        """
+        Inits class by loading the clustering result files
+        """
+        if isinstance(source, str):
+            with open(source) as f:
+                self.sign_to_pred_cluster = json.load(f)
+        else:
+            self.sign_to_pred_cluster = source
+            
+    def predict(
+        self, 
+        block_dict: Dict[str, List[str]], 
+        dataset: ANDData, 
+        use_s2_clusters: bool
+    ) -> Dict[str, List[str]]:
+        """
+        Creates dict that maps cluster name to the signatures it owns
+        """    
+        pred_clusters = {}
+        uknown_id = 0
+        for signatures in block_dict.values():
+            for signature in signatures:
+                if signature not in self.sign_to_pred_cluster:
+                    pred_clusters[f'uknown_{uknown_id}'] = [signature]
+                    uknown_id += 1
+                else:
+                    cluster = self.sign_to_pred_cluster[signature]
+                    if cluster not in pred_clusters:
```

Comment:
We could use defaultdict to avoid this if..else.

Reply:
True, I will make the change, thanks!

Reply:
Regarding param tuning of clusters, maybe you mean this: https://github.com/atypon/S2AND/blob/00d7c5c34b7baf6d888ec31573ef83c018c9ce93/src/run_clustering.py#L23 ?

Reply:
Oh yeah, sorry many changed files I was lost :P

Reply:
Check out the __call__ of the objective. We were discussing this for optuna optimization without the need of declaring the function in the same script using global variables.